### DISCUSSION TITLE
Event processing for map storage factories design discussion

### DISCUSSION THREAD

**Discussion by sguilhen** [keycloak/keycloak#9764] [Archive]
Provider factories can register as `ProviderEventListener`s with the `KeycloakSessionFactory`, which allows them to respond to events such as `RoleRemovedEvent` or `UserRemovedEvent`.

In the storage layer, the logic that is currently implemented in the legacy `JPA` provider factories can't be simply copied to the new map storage factories as the former factories make assumptions about the storage structure that may not be true in the new storage configuration.

Ideally the map storage factories should "translate" the provider event, that contains references to model classes, into events that are appropriate for the `MapStorage`s, and propagate these events into the storages that are meant to be handling the event (instead of using model classes, the factory can relay, for example, a `remove-role` event with a `ModelCriteria` that the storage should use to locate the role to be removed).

The context in which the event was generated is important, as we can have different storages at the `realm` level, so an event generated in the context of a specific `realm` should be handled by the corresponding map storage.

This thread aims to start a design discussion for event handling by the map storage factories.

**Comment by hmlnarik** [comment]
This has been now implemented by preferring `InvalidationHandler` over events and leveraging it in #11076.
